src/serialization/sb3.js

```diff
@@ -25,14 +27,156 @@ const {deserializeCostume, deserializeSound} = require('./deserialize-assets.js'
+ *
+ * @param {object} inputs The inputs to serialize
+ * @return {object} An object representing the serialized inputs
+ */
+const serializeInputs = function (inputs) {
+    const obj = Object.create(null);
+    for (const inputName in inputs) {
+        if (!inputs.hasOwnProperty(inputName)) continue;
+        // if block and shadow refer to the same block, only serialize one
+        if (inputs[inputName].block === inputs[inputName].shadow) {
+            // has block and shadow, and they are the same
+            obj[inputName] = [
+                INPUT_SAME_BLOCK_SHADOW,
+                inputs[inputName].block
+            ];
+        } else if (inputs[inputName].shadow === null) {
+            // does not have shadow
+            obj[inputName] = [
+                INPUT_BLOCK_NO_SHADOW,
+                inputs[inputName].block
+            ];
+        } else {
+            // block and shadow are both present and are different
+            obj[inputName] = [
+                INPUT_DIFF_BLOCK_SHADOW,
+                inputs[inputName].block,
+                inputs[inputName].shadow
+            ];
+        }
+    }
+    return obj;
+};
+
+/**
+ * Serialize the fields of a block in a more compact form.
+ * @param {object} fields The fields object to serialize
+ * @return {object} An object representing the serialized fields
+ */
+const serializeFields = function (fields) {
+    const obj = Object.create(null);
+    for (const fieldName in fields) {
+        if (!fields.hasOwnProperty(fieldName)) continue;
+        obj[fieldName] = [fields[fieldName].value];
+        if (fields[fieldName].hasOwnProperty('id')) {
+            obj[fieldName].push(fields[fieldName].id);
+        }
+    }
+    return obj;
+};
+
+/**
+ * Serialize the given block in the SB3 format with some compression of inputs,
+ * fields, and primitives.
+ * @param {object} block The block to serialize
+ * @return {object | array} A serialized representation of the block. This is an
+ * array if the block is one of the primitive types described above or an object,
+ * if not.
+ */
 const serializeBlock = function (block) {
+    const serializedPrimitive = serializePrimitiveBlock(block);
+    if (serializedPrimitive) return serializedPrimitive;
+    // If serializedPrimitive is null, proceed with serializing a non-primitive block
     const obj = Object.create(null);
-    obj.id = block.id;
     obj.opcode = block.opcode;
+    // NOTE: this is extremely important to serialize even if null;
+    // not serializing `next: null` results in strange behavior with block
+    // execution
     obj.next = block.next;
     obj.parent = block.parent;
-    obj.inputs = block.inputs;
-    obj.fields = block.fields;
+    obj.inputs = serializeInputs(block.inputs);
+    obj.fields = serializeFields(block.fields);
     obj.topLevel = block.topLevel ? block.topLevel : false;
     obj.shadow = block.shadow;
```

Comment:
There are a few fields here that seem like they could be reconstructed at load-time. For example, it seems like we should only need one of inputs / fields / shadow -- although it also seems likely that I don't fully understand what each of these is for ;). Anyway... is the plan to pare this down later as we have time to make deserialization do more work?

Reply:
That was not the plan, and that's largely due to our timeline, the current state of the VM/blocks/gui, and the nature of the Save/Load work in general (i.e. that it is a high risk feature, where once it is released we can only make changes that are strictly more permissive; supporting a plan to changing how we serialize fields and inputs in the future would necessitate supporting both versions of sb3 serialization during validation).

In my mind, this is part of the decision to have the little bit of pre-compression work as proposed in this PR, but other than that, avoid a lot of extra (and error-prone) work to piece together info that is currently still scattered across scratch-vm, scratch-blocks, and scratch-gui  about how each block is defined and what kinds of inputs and fields it expects in order to be able to reconstruct it properly during deserialization.

The rest of this post is an attempt to explain the differences between inputs, fields, and shadows, and what kind of information they represent. You probably know most if not all of this already given extension-ification work, but it helps to have it all written down in one place. 

### Inputs, Fields, and Shadows Galore

A block's 'inputs' property describes each of the arguments on the block, naming them, and what is currently connected to those inputs. A block's field describes the value of a named field in that block.

#### Side note: 
In both cases, these properties are named, as opposed to positioned in the order they appear on the block. Information about what input/field names can exist in a given block, in what order, is scattered in several places. The extension-ification work will make it so all of that information is unified and can be found in one place accessible using the block's opcode, but unfortunately to access that information without the extension-ification work would require hand writing something like the sb2_specmap, but one which maps sb3 opcodes to its argmap.

#### Example of inputs and fields:

For example, a broadcast block has one input with the key 'BROADCAST_INPUT'. When a broadcast block's input is unobscured, the `inputs` field on the block will look something like this (in the runtime, unserialized):

```inputs: {
    'BROADCAST_INPUT': {
        'name': 'BROADCAST_INPUT',
        'block': 'this is a uid for a shadow block which contains the broadcast dropdown menu'
        'shadow: 'this is a uid for a shadow block which contains the broadcast dropdown menu'
    }
}
```

#### Block vs. Shadow in an Input

The fact that block and shadow in the example above have the same uid indicates that there the broadcast block has a pluggable input and that there is nothing plugged into the input.

- If this was a block without a droppable field, the 'shadow' property would be null.

- If there was something plugged in, the 'block' and 'shadow' properties would have different uids. The block uid would refer to whatever was plugged in, and the shadow in both this new case and the case in the example above refers to the shadow block which contains the drop down menu as a *field* (described below).

#### Fields

The shadow block with the uid `this is a uid for a shadow block which contains the broadcast dropdown menu` has opcode 'event_broadcast_menu' and has a fields property which contains a single field describing the currently selected value in the dropdown menu. It looks like the following:

```fields: {
        'BROADCAST_OPTION': {
            'name': 'BROADCAST_OPTION',
            'value': 'this is the name of the broadcast message I have selected in the dropdown menu',
            'variableType': 'broadcast_msg',
            'id': 'this is the uid for the variable that represents this broadcast message'
    }
}
```

This field, because it refers to a broadcast menu dropdown has special properties like id and variableType required for connecting the selected value to the correct broadcast message variable). The field in a 'math_number' block for example would have just the  'name' and 'value' properties.

In the example above, having both the input and the field are important, because the input describes the current state of the inputs on the block, and allows us to preserve the contents of underlying shadows e.g. what was selected (via a dropdown) or typed into a field before the user plugged something into it.

This was really long, but I hope it helps to clarify some of the concerns and also what information is available where.

Reply:
Thanks - that's actually very helpful! Most of my extension work focuses on defining a block at a time, in the abstract: I usually don't have to worry about blocks which are connected to each other or even instantiated. Basically, if there's a UID involved then I haven't worked with it much ;)

Reply:
ah, gotcha! That makes sense.